engineering/architecture/dateandtime.md

```diff
@@ -0,0 +1,107 @@
+---
+title: Human Readable Dates and Times
+description: Spec for handling dates and times in a human friendly way.
+published: true
+date: 2022-02-14T12:35:23.307Z
+tags: 
+editor: markdown
+dateCreated: 2022-02-15T14:58:57.900Z
+---
+
+## Goals
+
+1. Users should be able to interact with dates, times, datetimes, and time durations in Mathesar in the most human-friendly way possible
+2. The UI should be extremely forgiving w.r.t. date/time input, and also display dates and times in a human-friendly way.
+3. The API should be extremely forgiving w.r.t. date/time input, but should return dates, times, and time durations in a more programmatic, i.e., consistent way than the UI.
+4. The UI and API should be consistent in the following way:  If the input string is understood by both UI and API, they should always result in the _same date/time_ being stored, regardless of the input method.
+5. The dates/times understood by the API should be a subset of those understood by the UI. That is, there should be no dates/times/time durations that are understood by the API, but which can't be entered through the UI.
+
+## Overall input flow
+
+If the UI is able to parse a date, time, duration, etc., it should reformat it into the appropriate canonical form described below and send that as the input to be stored to the API. If it's not able to parse the input, it should send the raw input string to the API and let the service layer / database try to parse the string. This helps satisfy goal number 5. If the service layer is able to parse the string, then store it. Otherwise, raise a validation error.
```

Comment:
I don't think the frontend should be compelled to send raw input to the API if it can't parse it, and seems unnecessary given goal 5. Someone might enter "pizza" in a datetime field, we don't need the backend to know that's not a datetime.

Reply:
After reading @pavish's [comment](https://github.com/centerofci/mathesar-wiki/pull/35#discussion_r809144659) and thinking about it more, it seems like the frontend wants to rely entirely on the backend for validation. I think this is fine for now, which makes my comment above irrelevant. 

I do think we should eventually look into whether we can do validation on the frontend (to make the UX snappier), but we can update this spec when it comes to that.

Reply:
My current thinking is to do the "default" parsing in the DB (and therefore for the API), and then the front end could extend that bit by bit, and add validation iteratively. But if we assume that the DB will need to attempt parsing whatever it's handed from the beginning, it takes pressure off of adding any intelligence in the front end for awhile.

Reply:
Sounds good.